### PR TITLE
Feature/us656983 refactor code

### PR DESCRIPTION
**Pull request by uppoju**
Refactoring code for better management

### CODE REVIEW

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java [hunk 1]
```diff
@@ -156,4 +162,18 @@ public void sortArgumentsAndResults() {
         setArguments(getArguments().stream().collect(toCollection(() -> new TreeSet<>(Comparator.comparing(EncassArgument::getName)))));
         setResults(getResults().stream().collect(toCollection(() -> new TreeSet<>(Comparator.comparing(EncassResult::getName)))));
     }
+
+    @Override
+    public boolean hasBundleAnnotation(){
+        if(hasAnnotated()){
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java:168]: This `hasAnnotated()` should be changed to `hasAnnotation()` or `isAnnotated()`.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java [hunk 2]
```diff
@@ -156,4 +162,18 @@ public void sortArgumentsAndResults() {
...
+            return annotations.stream().anyMatch(annotation -> ANNOTATION_TYPE_BUNDLE.equals(annotation.getType()));
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java:169]: This logic can be changed to: `annotations.contains(new Annotation(ANNOTATION_TYPE_BUNDLE));` as it is a `Set` and `Annotation.equals()` look only for type. Will reduce the complexity to constant time O(1).

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java [hunk 3]
```diff
@@ -156,4 +160,34 @@ public void sortArgumentsAndResults() {
         setArguments(getArguments().stream().collect(toCollection(() -> new TreeSet<>(Comparator.comparing(EncassArgument::getName)))));
         setResults(getResults().stream().collect(toCollection(() -> new TreeSet<>(Comparator.comparing(EncassResult::getName)))));
     }
+
+    @Override
+    public AnnotatedEntity<GatewayEntity> getAnnotatedEntity(final String projectName,
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java:165]: **Avoid taking project name and version**. You might earn something down the line. But, from design, why would I inject project details while getting the annotated entity.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java [hunk 4]
```diff
@@ -33,7 +37,8 @@
 @Named("ENCAPSULATED_ASSERTION")
 @ConfigurationFile(name = "encass", type = JSON_YAML)
 @EnvironmentType("ENCAPSULATED_ASSERTION")
-public class Encass extends GatewayEntity {
+@BundleGeneration
+public class Encass extends GatewayEntity implements AnnotatableEntity {
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java:41]: I like this idea.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java [hunk 5]
```diff
@@ -156,4 +160,34 @@ public void sortArgumentsAndResults() {
...
     }
+
+    @Override
+    public AnnotatedEntity<GatewayEntity> getAnnotatedEntity(final String projectName,
+                                                             final String projectVersion) {
+        return annotatedEntity == null ? createAnnotatedEntity(annotations, projectName, projectVersion) :
+                annotatedEntity;
+    }
+
+    public void populateBundleInfo(final AnnotatedEntity<GatewayEntity> annotatedEntity, final Annotation bundleAnnotation, final String projectName,
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java:171]: I suggest you to move this logic into AnnotatedEntity.
Once you get the annotated entity, you will call some method where you inject additional details like project name, version, etc.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/EntityUtils.java [hunk 6]
```diff
@@ -109,5 +123,10 @@ public FileType getFileType() {
         public String getEnvironmentType() {
             return environmentType;
         }
+
+        public boolean isBundleGenerationSupported() {
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/EntityUtils.java:127]: Why can't you silently compute this value inside?
You have entityClass property to do so.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/EntityUtils.java [hunk 7]
```diff
@@ -78,15 +89,18 @@ private EntityUtils() {
         private FileType fileType;
         private String environmentType;
 
+        private boolean bundleGenerationSupported;
+
         private GatewayEntityInfo() {
         }
 
-        private GatewayEntityInfo(String type, Class<? extends GatewayEntity> entityClass, String fileName, FileType fileType, String environmentType) {
+        private GatewayEntityInfo(String type, Class<? extends GatewayEntity> entityClass, String fileName, FileType fileType, String environmentType, boolean bundleGenerationSupported) {
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/EntityUtils.java:97]: I suppose bundleGenerationSupported is computable right inside. 
No need to make it as constructor-level arg.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/AnnotatedEntity.java [hunk 8]
```diff
@@ -65,36 +92,36 @@ public String getDescription() {
...
 
-    public void setRedeployableType(boolean isRedeployableType) {
-        this.isRedeployableType = isRedeployableType;
+    public void setRedeployable(boolean isRedeployableType) {
+        this.redeployable = isRedeployableType;
     }
 
-    public boolean isExcludeTypeEnabled() {
-        return isExcludeType;
+    public boolean exclude() {
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/AnnotatedEntity.java:119]: rename it to isExcluded

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/linker/ClusterPropertyLinker.java [hunk 9]
```diff
@@ -43,4 +45,9 @@ public void link(Bundle filteredBundle, Bundle bundle, File rootFolder) {
         idsToRemove.forEach(clusterPropertyMap::remove);
     }
 
+    @Override
+    public void link(Set<Dependency> dependencies) {
```
Comment (graju256) [inline gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/linker/ClusterPropertyLinker.java:49]: Can you please explain how does it work?
Why are you changing the type from CW to ENCAS?

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/linker/EntitiesLinker.java [hunk 10]
```diff
@@ -17,4 +19,7 @@
     default void link(Bundle filteredBundle, Bundle bundle, File rootFolder) {
         link(filteredBundle, bundle);
     }
+
```
Comment (graju256) [inline gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/linker/EntitiesLinker.java:22]: Please provide method level comment to understand why we need this and when does it get called?

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/paths/PathUtils.java [hunk 11]
```diff
@@ -40,4 +40,14 @@ public static String unixPathEndingWithSeparator(@NotNull String first, String..
     public static String unixPathEndingWithSeparator(@NotNull Path path) {
         return separatorsToUnix(path.toString()) + UNIX_PATH_SEPARATOR;
     }
+
+    public static String extractName(String nameWithPath) {
+        int index = nameWithPath.lastIndexOf(UNIX_PATH_SEPARATOR);
+        return index > -1 ? nameWithPath.substring(index + 1) : nameWithPath;
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/paths/PathUtils.java:46]: Interesting. I've never used condition like above. I always go with the != w.r.y. -1. 
Never mind.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BundleElementNames.java [hunk 12]
```diff
@@ -25,6 +25,7 @@
     public static final String ATTRIBUTE_TAG = "tag";
     public static final String ATTRIBUTE_ROOT_URL = "rootUrl";
     public static final String ATTRIBUTE_SOURCE_URL = "sourceUrl";
+    public static final String ATTRIBUTE_STRING_VALUE="stringValue";
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/gateway/BundleElementNames.java:28]: Space is missing in separating the "=" operator from the operands.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/config/spec/BundleGeneration.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * Specifies if bundle can be generated.
+ */
+@Documented
+@Retention(RUNTIME)
+@Target(TYPE)
+public @interface BundleGeneration {
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/config/spec/BundleGeneration.java:22]: It is almost like @bundle annotation.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/config/loader/PolicyAndFolderLoader.java [hunk 14]
```diff
@@ -66,11 +62,14 @@ private void loadPoliciesMetadata(final File rootDir, final Map<String, Policy>
                 policyDependencyMap.put(new Dependency(policyMetadata.getName(), EntityTypes.POLICY_TYPE),
                         new LinkedList<>(dependencies));
 
-                //Policy policy = policies.get(policyMetadata.getFullPath());
-                //policy.setAnnotations(policyMetadata.getAnnotations());
+                Policy policy = policies.get(policyMetadata.getFullPath());
```
Comment (graju256) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/config/loader/PolicyAndFolderLoader.java:65]: Finally, I could see policy details at one place.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java (hunk 3 above)
Comment (uppoju) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/Encass.java:165]: yeah, i felt the same, actually we can create annotated entity when loading the entity from yml files as it is design time info

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/linker/ClusterPropertyLinker.java (hunk 9 above)
Comment (uppoju) [reply gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/linker/ClusterPropertyLinker.java:49]: all cluster wide properties are changed to global environment properties when writing it to yml files
as a result when matching dependency with entities in the bundle its failing to load
So we are changing this in policy dependencies also

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/EntityUtils.java (hunk 7 above)
Comment (uppoju) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/EntityUtils.java:97]: done

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/EntityUtils.java (hunk 6 above)
Comment (uppoju) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/EntityUtils.java:127]: done

**Pull request merged**